# Prepare data

```{r prepare_data_libraries}
library(tidyverse)
```

## Genome metadata

```{r prepare_genome_metadata, warning=FALSE, comments="", message=FALSE}
#Completeness, contamination, length
genome_info <- read_csv("resources/genomeInfo.csv") %>%
  select(-N50) %>%
  mutate(genome = str_remove(genome, "\\.fa$"))

# Taxonomy
genome_taxonomy <- read_tsv("resources/gtdbtk.bac120.summary.tsv") %>%
  dplyr::rename(genome=1) %>%
  mutate(classification = str_replace_all(classification, ".__", "")) %>%
  separate(col = classification, sep = ";", into = c("domain", "phylum", "class", "order", "family", "genus", "species")) %>% 
  select(1:8) 

genome_metadata <- inner_join(genome_taxonomy, genome_info, by="genome")
```

## Read counts

```{r load_reads, warning=FALSE, comments="", message=FALSE}
read_counts <- read_tsv("resources/coverm_read_counts.tsv")
```

## Coverage calculation

```{r genome_coverage, warning=FALSE, comments="", message=FALSE}
basehits <- read_tsv("resources/coverm_covered_bases.tsv") %>%
    dplyr::rename(genome=1) %>%
    arrange(match(genome, read_counts$genome))

genome_coverage <- basehits %>%
  mutate(across(where(is.numeric), ~ ./genome_metadata$length))
```

## Coverage filtering

```{r genome_filtering, warning=FALSE, comments="", message=FALSE}
min_coverage=0.3

read_counts_filt <- genome_coverage %>%
  mutate(across(where(is.numeric), ~ ifelse(. > min_coverage, 1, 0))) %>%
  mutate(across(-1, ~ . * read_counts[[cur_column()]])) 
```

## Read counts to genome counts

```{r genome_counts, warning=FALSE, comments="", message=FALSE}
readlength=150

genome_counts <- read_counts %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) ))

genome_counts_filt <- read_counts_filt %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) ))
```

## Genome tree


```{r load_genome_tree, warning=FALSE, comments="", message=FALSE}
genome_tree <- read_tree("resources/gtdbtk.backbone.bac120.classify.tree")

genome_tree <- keep.tip(genome_tree, tip=genome_metadata$genome)
```

## Prepare color scheme

```{r get_ehi_colors_skin, warning=FALSE, comments="", message=FALSE}
phylum_colors <- read_tsv(
  "https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
  mutate_at(vars(phylum), ~ str_replace(., "[dpcofgs]__", "")) %>% # Clean taxonomy prefix
  right_join(genome_metadata, by = join_by(phylum == phylum)) %>% # Keep only phyla in our dataset
  arrange(match(genome, genome_tree$tip.label)) %>% # Align colors with tree tips
  select(phylum, colors) %>%
  unique() %>%
  arrange(phylum) %>%
  pull(colors, name = phylum)
```

# Wrap working objects

```{r savedata, comment="", message=FALSE, warning=FALSE}
save(
  #sample_metadata,
  genome_metadata,
  genome_counts_filt,
#  preprocess_info,
  genome_tree,
#  genome_gifts,
#  microbial_fraction,
  phylum_colors,
  file = "resources/metagenomics.Rdata")
```


